crates/medea_macro/src/enum_delegate.rs

```diff
@@ -0,0 +1,100 @@
+//! `#[enum_delegate]` macro implementation.
+
+use std::iter;
+
+use proc_macro::TokenStream;
+use quote::quote;
+
+/// Generates the actual code for `#[enum_delegate]` macro.
+///
+/// # Generation algorithm
+/// 1. parse input enum;
+/// 2. check that enum is not empty;
+/// 3. add "{}" to the macro argument to get the full function declaration;
+/// 4. parse this full function declaration;
+/// 5. check that function to delegate is not static by counting;
+///    `[&][&mut] self` arguments;
+/// 6. get all argument idents (except `[&][&mut] self`) from a function to be
+///    delegated;
+/// 7. generate function with `match` of all variants that returns
+///    result of delegated function call;
+/// 8. add this function to implementation of enum.
+pub fn derive(args: &TokenStream, input: TokenStream) -> TokenStream {
+    let mut output = input.clone();
+    let inp: syn::DeriveInput =
+        syn::parse(input).expect("failed to parse input");
+
+    let mut enum_name_iter = iter::repeat(inp.ident);
+    let enum_name = enum_name_iter.next().unwrap();
+
+    let variants: Vec<syn::Ident> = match inp.data {
+        syn::Data::Enum(data) => {
+            data.variants.into_iter().map(|c| c.ident).collect()
+        }
+        _ => panic!("This macro should be used only with enums!"),
```

Comment:
Let's not use panic-oriented code. Using [`syn::parse::Error`](https://docs.rs/syn/0.15.34/syn/parse/struct.Error.html) is more idiomatic here as allows to specify the place in code where shit happens (`span`s).